### ISSUE TITLE
[1.6.1] Upgrade from 1.6.0.1 to 1.6.1 broke cleaner library

### ISSUE THREAD

**Issue by artgo** [meteor/meteor#9630]
We are using cleaner library to clean DB before tests. It stopped working after 1.6.1 upgrade.
Here is the reference ticket: https://github.com/xolvio/cleaner/issues/20

**Comment by benjamn** [comment]
I can't tell from that issue what the problem is. Can you provide more details (either here or there)?

**Comment by artgo** [comment]
@nikolai-berezin can you please help?

**Comment by nikolai-berezin** [comment]
I found that error fires after calling fiber.yield() function in cleaner library. 
After that the next error rises:
`Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.`
I've tried to increase timeout and call done() function. Nothing helps.

**Comment by charlesdeb** [comment]
@benjamn I can't speak for the other guys in this thread, but my issue is that calling `resetDatabase()` which is provided by the `xolvio:cleaner` packager, in my `beforeEach` hooks in tests results in the same timeout issue that @artgo and @nikolai-berezin have flagged. There is no stack trace - just the timeout error as mentioned above. I have had to switch out `practicalmeteor:mocha` to `cultofcoders:mocha` as well since there is some coffeescript dependency fail. It appears to have been about two years since the `xolvio:cleaner` and `practicalmeteor:mocha` had any work done on them.

**Comment by brookback** [comment]
We're getting the same error on Meteor 1.6.1 with the `cleaner` package. This is the test output:

```
I20180208-16:57:07.459(1)?   Namespace
I20180208-16:57:07.460(1)?     publication
=> Meteor server restarted
      1) "before each" hook: wrappedFunction for "should publish"
I20180208-16:57:09.470(1)?
I20180208-16:57:09.470(1)?
I20180208-16:57:09.471(1)?   0 passing (2s)
I20180208-16:57:09.471(1)?   1 failing
I20180208-16:57:09.471(1)?
I20180208-16:57:09.471(1)?   1) publication "before each" hook: wrappedFunction for "should publish":
I20180208-16:57:09.472(1)?      Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.
```

The test code includes:

```js
import { resetDatabase } from 'meteor/xolvio:cleaner';

beforeEach(() => {
    resetDatabase();
  });
});
```

The business logic and test code was untouched in the upgrade to 1.6.1.

The code in the `cleaner` package is quite short: https://github.com/xolvio/cleaner/blob/master/cleaner.js

Seems they're using `MongoInternals.defaultRemoteCollectionDriver().mongo.db` and `Meteor.wrapAsync` to remove all documents from a collection. Perhaps have the former changed in 1.6.1? Also, we've got [`Collection#rawCollection`](https://docs.meteor.com/api/collections.html#Mongo-Collection-rawCollection) now instead.

**Comment by brookback** [comment]
Ok, I found it: changing line 26 to

```js
remove({}, {});
```

makes it work 🎉 The second param symbolises an options object.

I did a `console.log` of `remove.toString()` in the cleaner package, and it turns out the function signature is:

```
remove = (selector, options, callback)
```

If `typeof callback === 'function'`, the function will call it. In our case, `Meteor.wrapAsync` made the last parameter, `options`, become a callback, and the "real" callback parameter was `undefined`.

Either way, I don't know if this behaviour changed in Meteor 1.6.1, but passing a second param to `remove` fixes it.

**Comment by charlesdeb** [comment]
That's great @brookback. Do I just need to fix this in my local node_modules folder then? I'm not sure if the xolvio guys would do anything with a PR even if one was to be sent...

**Comment by nikolai-berezin** [comment]
@brookback Thanks for your investigation. Seems like creating a PR on xolvio is not an option cause it's not longer supported by developers. Looking for a better solution to clean a db on running tests. 
How you guys makes a db reset in your tests?

**Comment by charlesdeb** [comment]
@benjamn, any suggestions on the best way to proceed with this? It would seem that the testing [tutorial](https://guide.meteor.com/testing.html#test-data) on the meteor docs site is now out of date since it still recommends `xolvio:cleaner`, and - for some users at least - it no longer works for 1.6.1.

**Comment by brookback** [comment]
@charlesdeb @nikolai-berezin The beauty of Meteor packages is that you simply can clone them down into your local `packages` dir:

```
git clone git@github.com:xolvio/cleaner.git packages/cleaner
```

And then apply a fix in there. Since the package itself doesn't get many updates these days, it's not a biggie (to us) to keep a local, embedded copy of it.

The other option is of course to fork and publish a maintained version on Atmosphere 😄 

But @charlesdeb has a point regarding the official recommendation of `xolvio:cleaner` in the Meteor Guide. If it breaks on the current Meteor version, MDG should recommend a working fork instead.

**Comment by hwillson** [comment]
As @brookback mentioned in https://github.com/meteor/meteor/issues/9630#issuecomment-364851969, this can be fixed by modifying a local clone/package of `cleaner`, or by forking, fixing and maintaining a new official version.

As for updating the Guide, I agree. We'll track this in https://github.com/meteor/guide/issues/705. Thanks!

**Comment by lgandecki** [comment]
We applied the @brookback fix in 0.3.3 , I verified it in the freshly cloned todos app by putting:

```javascript

Meteor.methods({
  'resetDb': () => {
    resetDatabase()
  }
})
```

on the server side and calling it from the client. 

BTW. Why do you guys think xolv.io wouldn't take a PR? We are still here :-) The package hasn't been updated in a while but it did its job and didn't really require any updates till now. Is this a general feeling about xolv.io meteor-related work or about this particular page? 
In other words - do you guys have any other outstanding issues we could help with?

Thanks!

**Comment by brookback** [comment]
@lgandecki Thanks! GJ. I think the package indeed just "does its job", and I personally think there's an overall, incorrect decision algorithm in the open source community that says

```
if last_commit.date > 6.months.ago
   mentallyMarkAsUnmaintained()
```

I think packages, like yours, which do their job and solves a particular problem are perfectly fine without updates, apart from fixes like this one. "New code, new bugs!" ;)

Keep it up! It's a solid package.

**Comment by nikolai-berezin** [comment]
@lgandecki Thank you guys for keeping your package up to date. Great job. Keep up!
